crates/core/src/operations/connect.rs

```diff
@@ -191,34 +216,32 @@ impl RelayState {
         &mut self,
         ctx: &C,
         observed_remote: &PeerKeyLocation,
-        observed_addr: SocketAddr,
     ) -> RelayActions {
         let mut actions = RelayActions::default();
         push_unique_peer(&mut self.request.visited, observed_remote.clone());
         push_unique_peer(&mut self.request.visited, ctx.self_location().clone());
 
-        if self.request.origin.peer.addr.ip().is_unspecified()
-            && !self.observed_sent
-            && observed_remote.peer.pub_key == self.request.origin.peer.pub_key
-        {
-            self.request.origin.peer.addr = observed_addr;
-            if self.request.origin.location.is_none() {
-                self.request.origin.location = Some(Location::from_address(&observed_addr));
+        if let Some(joiner_addr) = self.request.observed_addr {
+            if !self.observed_sent && !is_publicly_routable(self.request.joiner.peer.addr.ip()) {
```

Comment:
Why do we need the is_publically_routable function still?

Reply:
Switched to always trusting the observed socket and removed the helper. The observed UDP source address is ground truth - it provably works for return traffic. Even in multi-homed/NAT scenarios, we can'\t verify any other advertised address is reachable.